index.d.ts

```diff
@@ -1,47 +1,49 @@
 // Project: Immutability helper
-// TypeScript Version: 2.2
+// TypeScript Version: 2.8
 
 export default update
 
-declare function update<T>(
-  data: ReadonlyArray<T>,
-  query: ArrayOperators<T>,
-): ReadonlyArray<T>
-
-declare function update<T>(
-  data: ReadonlySet<T>,
-  query: SetOperators<T>,
-): ReadonlySet<T>
-
-declare function update<K, V>(
-  data: ReadonlyMap<K, V>,
-  query: MapOperators<K, V>,
-): ReadonlyMap<K, V>
-
-declare function update<T>(data: T, query: Query<T>): T
-
-type Tree<T> = {[K in keyof T]?: Query<T[K]>}
-export type Query<T> =
-  | Tree<T>
-  | ObjectOperators<T>
-  | ArrayOperators<any>
-  | SetOperators<any>
-
-type ObjectOperators<T> =
-  | {$set: any}
-  | {$toggle: Array<keyof T | number>}
-  | {$unset: Array<keyof T | number>}
-  | {$merge: Partial<T>}
-  | {$apply: (old: T) => T}
-  | ((old: T) => any)
-type ArrayOperators<T> =
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+declare function update<T>(target: T, spec: Spec<T>): T
+
+export type Spec<T> =
+  T extends (Array<infer U> | ReadonlyArray<infer U>) ? ArraySpec<U> :
+  T extends (Map<infer K, infer V> | ReadonlyMap<infer K, infer V>) ? MapSpec<K, V> :
+  T extends (Set<infer U> | ReadonlySet<infer U>) ? SetSpec<U> :
+  T extends object ? ObjectSpec<T> :
+  BaseSpec<T>
+
+type BaseSpec<T> =
+  | { $set: T }
+  | { $apply: (v: T) => T }
+  | ((v: T) => T)
```

Comment:
I tried putting something like the original custom command types right here:

```ts
| { [customCommand: string]: any }
```

But that basically turned off all type checking. That type matches pretty much everything, so if Typescript can't find a well-typed match, it'll pick this union member and then all bets are off. That's especially unfortunate since a huge part of the value of types here is to catch rename-refactor changing the shape of `T`, which this effectively disables.

I tried two methods to resolve this, but neither seemed to work. Perhaps I was turning the type system knobs incorrectly.

1.  Defining another type parameter the explicitly names the custom keys.

    ```ts
    declare function update<T, K extends PropertyKey = never>(target: T, spec: Spec<T, K>): T

    // Punching K around through `Spec` where necessary, then...

    type BaseSpec<T, K extends PropertyKey> =
      ...
      | { [customCommand: K]: any }
    ```

    Typescript didn't like this because `K` is not literally `string`, even if you made it so `K extends string`, so instead you can:

    ```ts
    declare function update<T, K extends string = never>(target: T, spec: Spec<T, K>): T

    type BaseSpec<T, K extends string> = // Has to be string
      ...
      | { [C in K]?: any }
    ```

    Which almost type checks, but if you punch it through to all the places (namely, `ObjectSpec`) where it needs to get passed, Typescript will start complaining that

    > Type alias 'Spec' circularly references itself.

    Which is surprising, since it can easily reference itself in `ObjectSpec<T>` with `{ [K in keyof T]?: Spec<T[K]> }`, but adding that extra `K` type parameter to `Spec` breaks it. Perhaps this is because that type parameter appears on the left side (as a key) rather than the right (as a value)?

2. Similar to above, but instead of using a type to express the keys, I used a type to express the whole set of well-typed operations that are legal.

    ```ts
    declare function update<T, C extends object = {}>(target: T, spec: Spec<T, C>): T

    type BaseSpec<T, C extends object> =
      ...
      | C
    ```

    This has the same problem as above. I actually prefer this method because it's more precise in allowing you to well-type your custom commands if you type-parameterize your `update` instance properly. I'm a little confused why this is illegal, since if you collapse out the noise you should get something like `Spec<T, C> = ... | C` which seems pretty legal. I'll have to dig into this one more.

An idea I haven't tried is to allow type-parameterizing the different types of operations independently (like `update<T, A extends ArraySpec<T> ...>`, etc.). This could get REAL tedious, but might work around the limitations of self-referentiality. There's a [Typescript PR in the works right now](https://github.com/Microsoft/TypeScript/pull/23696) that would make it more usable, too, by allowing you to just override the one you want (assuming it works at all).

As a design goal of these approaches, I was trying to get the general case (no customization) to work excellently out of the box, and then only if you want to customize it do you have to provide extra type information (seems reasonable, I think). That's why I really don't like the `{ [customCommand: string]: any }` method as an option.

Reply:
Interestingly, it looks like this might be a heuristic around mutually recursive types failing. You can strip the failing example down to this:


```ts
export type Spec<T, C extends object> = ObjectSpec<T, C>

type ObjectSpec<T, C extends object> =
  | { [K in keyof T]?: Spec<T[K], C> }
  | C
```

But that `Spec` type is just a passthrough alias, so you can inline it because it isn't doing anything useful:

```ts
export type Spec<T, C extends object> = 
  | { [K in keyof T]?: Spec<T[K], C> }
  | C
```

At which point, it compiles successfully. So then of course I tried to inline the entire type into one gargantuan whole:

```ts
export type Spec<T, C extends object> = 
  | (
      T extends (Array<infer U> | ReadonlyArray<infer U>) ? (
        | { $push: T }
        | { $unshift: T }
        // Technically splice is variadic, but we stop after the third shape because that usage is very rare and we can't do variadic types here.
        | { $splice: Array<[number] | [number, number] | [number, number, T]> }
        // TODO: Wanted this to work...
        // | { [index: number]: Spec<T> }
        // TODO: This workaround doesn't seem to work though...
        // | { [K in keyof Array<T>]?: Spec<T> }
      ) :
      T extends (Map<infer K, infer V> | ReadonlyMap<infer K, infer V>) ? (
        | { $add: Array<[K, V]> }
        | { $remove: Array<K> }
      ) :
      T extends (Set<infer U> | ReadonlySet<infer U>) ? (
        | { $add: Array<T> }
        | { $remove: Array<T> }
      ) :
      T extends object ? (
        | { $toggle: Array<keyof T | number> }
        | { $unset: Array<keyof T | number> }
        | { $merge: Partial<T> }
        | { [K in keyof T]?: Spec<T[K], C> }
      ) : {}
    ) 
  | { $set: T }
  | { $apply: (v: T) => T }
  | ((v: T) => T)
  | C
```

But this doesn't compile, for the same reason. What gives?!